import-meta-path-helpers.md

```diff
@@ -0,0 +1,76 @@
+# Helpers on `import.meta` for working with paths
+
+## Background
+
+A common user request from developers writing ESM code is easy access to file path information of the current module. Or put another way, they miss the `__dirname` and `__filename` local variables from CommonJS, that define the directory and full path of the current module. These values are distinct from `import.meta.url`, which is the URL of the current module. To illustrate:
+
+- `import.meta.url` is a string like `file:///Users/Geoffrey/my-project/src/index.js`
+- `__filename` is a string like `/Users/Geoffrey/my-project/src/index.js`
+- `__dirname` is a string like `/Users/Geoffrey/my-project/src`
+
+Since `__filename` and `__dirname` have been around [since Node.js 0.1.27](https://nodejs.org/api/modules.html#__dirname), released in 2010, they are very well known to Node.js developers and very broadly referenced by `npm` packages. It is common to find documentation suggesting passing `__dirname` into a function for serving the static assets of a folder, for example, or passing `__filename` into a function used for logging.
+
+In Node.js, developers working with such path-based APIs need to jump through a few hoops. Node.js provides a helper `fileURLToPath` that developers can import from `node:url`, to enable code like `serveStaticAssets(dirname(fileURLToPath(import.meta.url)))`. A common complaint of Node.js developers is that this is too verbose, and a way to achieve the same with simpler code and without needing to import utility helpers would be preferable.
+
+## Proposal
+
+For runtimes that support filesystem-based modules, we propose adding two new optional properties to `import.meta`:
+
+### `import.meta.filename`
+
+A string containing the full absolute filesystem path to the current module, like CommonJS `__filename`. This is the same as `fileURLToPath(import.meta.url)` in Node.js. An example value would be `/Users/Geoffrey/my-project/src/index.js`.
+
+The value will be undefined for modules that are not loaded from the filesystem, such as modules whose `import.meta.url` begins with `http` or `data`.
```

Comment:
Conversation above was marked as resolved, but I do think it would be good to be explicit about whether in this scenario the property should present with the value `undefined`, or should be absent. I can't tell from this description, since `import.meta.filename === undefined` would be true either way.

(Personally I would expect it to be absent, although "present with value `undefined`" would allow distinguishing "this module was loaded not from the filesystem" from "this module was loaded in an environment that does not support `import.meta.filename`.)

Reply:
> although "present with value `undefined`" would allow distinguishing "this module was loaded not from the filesystem" from "this module was loaded in an environment that does not support `import.meta.filename`

I don't see a usecase where that would be useful. In all likeliness browsers will never define `filename` or `dirname`, so it makes sense to align I think.

```suggestion
The value will be absent for modules that are not loaded from the filesystem, such as modules whose `import.meta.url` begins with `http` or `data`.
```
or
```suggestion
The value will not be defined for modules that are not loaded from the filesystem, such as modules whose `import.meta.url` begins with `http` or `data`.
```

Reply:
If you want to be maximally clear, "the property will not be present" is the least ambiguous, I think. Doesn't make sense to speak about a "value" at all when the point is that the property which would hold the value doesn't exist.

Reply:
This was discussed in today’s WinterCG meeting and the consensus seemed to be that the property should be absent. This would mean there are only two potential states: property is present and has a value, or it’s not present. There is no third “property is present but is undefined” state for code to need to potentially deal with. This means that code can’t feature-detect whether the runtime supports this property in general but just not for the current module, but that ability doesn’t seem worth the additional complexity.